test/unit/org/apache/cassandra/auth/AuthCacheInvalidationTest.java

```diff
@@ -63,146 +83,161 @@ public void prepare() throws Throwable
-        executeNet("CREATE ROLE " + role1.getRoleName());
-        executeNet("CREATE ROLE " + role2.getRoleName());
+        createRole(role1);
+        createRole(role2);
         executeNet("GRANT " + role1.getRoleName() + " TO " + user1.getRoleName());
         executeNet("GRANT " + role2.getRoleName() + " TO " + role1.getRoleName());
 
         execute("CREATE KEYSPACE " + ks + " WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1}");
         execute("CREATE TABLE " + ks + '.' + tab + " (id int PRIMARY KEY)");
         execute("CREATE TABLE " + ks + '.' + tab2 + " (id int PRIMARY KEY)");
 
-        executeNet("GRANT SELECT, UPDATE ON TABLE " + ks + '.' + tab + " TO " + role1.getRoleName());
-        executeNet("GRANT UPDATE ON KEYSPACE " + ks + " TO " + role2.getRoleName());
+        executeNet("GRANT SELECT ON TABLE " + ks + '.' + tab + " TO " + role1.getRoleName());
+        executeNet("GRANT MODIFY ON TABLE " + ks + '.' + tab + " TO " + role1.getRoleName());
+        executeNet("GRANT MODIFY ON KEYSPACE " + ks + " TO " + role2.getRoleName());
 
-        DatabaseDescriptor.getAuthManager().invalidateCaches();
+        Roles.clearCache();
+        AuthenticatedUser.clearCache();
+    }
+
+    private void createRole(RoleResource role) throws Throwable
+    {
+        executeNet(format("CREATE ROLE %s WITH LOGIN = true AND password = '%s'", role.getRoleName(), userPw));
     }
 
     @Test
     public void testInvalidationOnGrant() throws Throwable
     {
+        populateAndCheckCaches();
+        
         // Checks that the user has no SELECT permissions on tab2
-
-        useUser(user1.getRoleName(), userPw);
-
-        assertUnauthorized("SELECT * FROM " + ks + '.' + tab2,
-                           "User " + user1.getRoleName() + " has no SELECT permission on <table " + ks + '.' + tab2 + "> or any of its parents");
+        assertSelectNotAllowed(user1, ks, tab2);
+        assertSelectNotAllowed(role1, ks, tab2);
+        assertSelectNotAllowed(role2, ks, tab2);
 
         // Checks that the caches have been populated by the query.
         assertRolesCacheContains(user1, setOf(role1));
         assertRolesCacheContains(role1, setOf(role2));
         assertRolesCacheContains(role2, setOf());
 
-        assertPermissionsCacheContains(user1, emptyMap());
-        assertPermissionsCacheContains(role1, permissionPerResource(DataResource.table(ks, tab), SELECT, UPDATE));
-        assertPermissionsCacheContains(role2, permissionPerResource(DataResource.keyspace(ks), UPDATE));
+        assertPermissionsCacheContains(user1, emptyPermissionPerResource(DataResource.table(ks, tab2)));
+        assertPermissionsCacheContains(user1, permissionPerResource(DataResource.table(ks, tab), SELECT, MODIFY));
+        assertPermissionsCacheContains(role1, permissionPerResource(DataResource.table(ks, tab), SELECT, MODIFY));
+        assertPermissionsCacheContains(role2, permissionPerResource(DataResource.keyspace(ks), MODIFY));
```

Comment:
In DSE there were no permissions cache entries for user1 (the emptyMap check). However, in C* there are entries, just none for tab2 in this case.

This kind of illustrates the uncertainty I have about the permissions cache checks currently; the DSE check for emptyMap is  complete - there are no entries. 

In an attempt to be more complete, here it's checking user a) has no entries for tab2, and b) has specific entries for tab.

Looking at this again now, probably the role1 and role2 asserts should also include one that checks tab2 is not present.